Add sinusoidal positional embedding module

antimora opened the pull request:
## Pull Request Template

### Checklist

- [x] Confirm that `run-checks.sh` has been executed.

NOTE: Currently the build on Mac M1 is broken, see https://github.com/burn-rs/burn/issues/480
so WGPU tests were skipped

### Changes

Add a brand new sinusoidal positional embedding module.

### Testing

Unit testing using values generated from a reference implementation.

nathanielsimard commented on burn-core/src/nn/sin_embedding.rs:
```diff
@@ -0,0 +1,198 @@
+    let inv_timescales = (0..d_model / 2)
+        .map(|x| expf(-log_timescale_increment * x as f32))
+        .collect::<Vec<_>>();
+
+    // The scaled_time is calculated by multiplying inv_timescales with an array of numbers
+    // from 0 to length - 1. This operation performs element-wise multiplication between
+    // each number in the sequence from 0 to length - 1 and the inv_timescales array.
+    // This results in a 2D array where each column is a sequence of numbers from 0 to length - 1,
+    // scaled by a different factor from inv_timescales.
+    let scaled_time = (0..length)
+        .map(|i| {
+            inv_timescales
+                .iter()
+                .map(|x| x * i as f32)
+                .collect::<Vec<_>>()
+        })
+        .collect::<Vec<_>>();
+
+    // Calculate the sine and cosine of each element in the scaled_time array,
+    // resulting in two arrays of the same shape.
+    let scaled_time_sin = scaled_time
+        .iter()
+        .map(|x| x.iter().map(|y| sinf(*y)).collect::<Vec<_>>())
+        .collect::<Vec<_>>();
+
+    let scaled_time_cos = scaled_time
+        .iter()
+        .map(|x| x.iter().map(|y| cosf(*y)).collect::<Vec<_>>())
+        .collect::<Vec<_>>();
+
+    // concatenate these two arrays along the second dimension
+    // (i.e., it adds the second array as new columns to the first array)
+    let scaled_time_sin_cos = scaled_time_sin
+        .into_iter()
+        .zip(scaled_time_cos.into_iter())
+        .map(|(mut x, mut y)| {
+            x.append(&mut y);
+            x
+        })
+        .collect::<Vec<_>>();
```
Would it be better to use tensor operations here instead? I think the arange function might be missing for the float tensor, but other than that it should be fine.

antimora replied:
It's easy to understand if the logic is in Rust and there is no advantage to use tensor ops for one time initialization. I'd suggest we'd leave it as is.

nathanielsimard commented on burn-core/src/nn/sin_embedding.rs:
```diff
@@ -0,0 +1,198 @@
+
+/// Lookup table to store sinusoidal vectors for positional embedding.
+#[derive(Module, Debug)]
+pub struct SinEmbedding<B: Backend> {
```
I would prefer `PositionalEncoding` instead of `SinEmbedding`.

nathanielsimard commented on burn-core/src/nn/sin_embedding.rs:
```diff
@@ -0,0 +1,198 @@
+#[derive(Config)]
+pub struct SinEmbeddingConfig {
+    /// The number of embedding vectors.
+    n_embedding: usize,
```
I would rename it to `max_sequence_size` since it's easier to know what value should be used. A default of 5000 is often common.

nathanielsimard commented on burn-core/src/nn/sin_embedding.rs:
```diff
@@ -0,0 +1,198 @@
+    /// - input: [batch_size, seq_length, d_model]
+    /// - output: [batch_size, seq_length, d_model]
+    pub fn forward(&self, input: Tensor<B, 3>) -> Tensor<B, 3> {
+        input.add(self.sinusoids.clone())
```
I think a slice is normally needed here. See https://pytorch.org/tutorials/beginner/transformer_tutorial.html for reference.

nathanielsimard replied:
I often see dropout as well

antimora commented:
I'll add the slice for varying length, and I will not add dropout per our discussiong. If someone wants it, they can add easily in their model.

antimora commented:
> It's easy to understand if the logic is in Rust and there is no advantage to use tensor ops for one time initialization. I'd suggest we'd leave it as is. 

I have refactored it to match positional encoding logic and simplified the logic.

antimora commented:
@nathanielsimard Fixed all the issues previously discussed. Ready for review again.
